File: Python/7-Trees/Problems/2-MaximumDepth-of-BinaryTree/MaximumDepth-of-BinaryTreePro.py
from typing import Optional
from collections import deque
import time
import pickle
import concurrent.futures
import json

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        """
        Initializes a TreeNode object.

        Args:
            val (int): The value of the node. Defaults to 0.
            left (Optional[TreeNode]): The left child of the node. Defaults to None.
            right (Optional[TreeNode]): The right child of the node. Defaults to None.
        """
        self.val = val
        self.left = left
        self.right = right

# Approach 1: Simulating Processing Time with Time Delays
def max_depth_with_delay(root: Optional[TreeNode]) -> int:
    """
    Calculates the maximum depth of a binary tree by recursively traversing the tree
    and adding a small time delay at each node to simulate processing time.  This is
    primarily for demonstration purposes to show the effect of processing delays.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        int: The maximum depth of the tree.
    """
    if not root:
        return 0  # Base case: if the node is None, the depth is 0
    time.sleep(0.05)  # Simulate delay in processing (e.g., I/O operation)
    # Recursively calculate the depth of the left and right subtrees, and return the greater depth + 1
    return 1 + max(max_depth_with_delay(root.left), max_depth_with_delay(root.right))

# Approach 2: Using Pickle Serialization
def max_depth_pickle(root: Optional[TreeNode]) -> int:
    """
    Calculates the maximum depth of a binary tree by serializing and deserializing
    the tree structure using pickle.  This approach is inefficient for depth calculation
    as it involves unnecessary object conversion.  It's included to demonstrate an alternative
    (but not recommended) approach.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        int: The maximum depth of the tree.
    """
    if not root:
        return 0  # Base case: if the node is None, the depth is 0
    serialized = pickle.dumps(root)  # Serialize the tree to a byte stream
    deserialized = pickle.loads(serialized)  # Deserialize the byte stream back to a tree
    # Recursively calculate the depth of the deserialized left and right subtrees
    return 1 + max(max_depth_pickle(deserialized.left), max_depth_pickle(deserialized.right))

# Approach 3: Multi-threaded Depth Calculation
def max_depth_multithreaded(root: Optional[TreeNode]) -> int:
    """
    Calculates the maximum depth of a binary tree using multiple threads. This approach
    attempts to improve performance by calculating the depths of the left and right
    subtrees concurrently. However, due to the Global Interpreter Lock (GIL) in Python,
    true parallelism is not achieved for CPU-bound tasks like this.  It might provide some
    benefit if the depth calculation involves some I/O, but for a pure in-memory
    tree traversal, the benefit is limited.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        int: The maximum depth of the tree.
    """
    if not root:
        return 0  # Base case: if the node is None, the depth is 0

    with concurrent.futures.ThreadPoolExecutor() as executor:
        # Submit the left and right subtree depth calculations to the thread pool
        left_future = executor.submit(max_depth_multithreaded, root.left)
        right_future = executor.submit(max_depth_multithreaded, root.right)
        # Get the results from the futures (this will block until the tasks are complete)
        left_depth = left_future.result()
        right_depth = right_future.result()

    return 1 + max(left_depth, right_depth)  # Calculate and return the maximum depth

# Approach 4: Using JSON Serialization
def max_depth_json(root: Optional[TreeNode]) -> int:
    """
    Calculates the maximum depth of a binary tree by serializing the tree to a JSON
    string and then deserializing it back into a tree structure.  This is extremely
    inefficient for calculating tree depth.  It's included as another example of an
    alternative approach, but it's not recommended for performance reasons.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        int: The maximum depth of the tree.
    """
    if not root:
        return 0  # Base case: if the node is None, the depth is 0
    json_data = json.dumps(tree_to_dict(root))  # Convert the tree to a JSON string
    root_dict = json.loads(json_data)  # Parse the JSON string back into a dictionary
    new_root = dict_to_tree(root_dict)  # Convert the dictionary back into a TreeNode
    return 1 + max(max_depth_json(new_root.left), max_depth_json(new_root.right))  # Recurse

# Helper function to convert tree to dictionary
def tree_to_dict(root: Optional[TreeNode]) -> dict:
    """
    Helper function to convert a binary tree to a dictionary representation.
    This is used for JSON serialization.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        dict: A dictionary representation of the tree.
    """
    if not root:
        return None  # Base case: if the node is None, return None
    # Recursively convert the node and its children to a dictionary
    return {"val": root.val, "left": tree_to_dict(root.left), "right": tree_to_dict(root.right)}

# Helper function to convert dictionary back to tree
def dict_to_tree(data: Optional[dict]) -> Optional[TreeNode]:
    """
    Helper function to convert a dictionary representation of a tree back to a binary tree.
    This is used for JSON deserialization.

    Args:
        data (Optional[dict]): The dictionary representation of the tree.

    Returns:
        Optional[TreeNode]: The root of the reconstructed binary tree.
    """
    if not data:
        return None  # Base case: if the data is None, return None
    # Recursively construct the tree from the dictionary
    return TreeNode(data["val"], dict_to_tree(data["left"]), dict_to_tree(data["right"]))

# Approach 5: BFS Using Queue
def max_depth_bfs(root: Optional[TreeNode]) -> int:
    """
    Calculates the maximum depth of a binary tree using Breadth-First Search (BFS) with a queue.
    This is an iterative approach and generally more efficient than recursive DFS for
    larger trees as it avoids the overhead of recursive function calls.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        int: The maximum depth of the tree.
    """
    if not root:
        return 0  # Base case: if the root is None, the depth is 0
    queue = deque([root])  # Initialize a queue with the root node
    depth = 0  # Initialize the depth counter
    while queue:
        depth += 1  # Increment the depth for each level of the tree
        for _ in range(len(queue)):  # Iterate through all nodes at the current level
            node = queue.popleft()  # Dequeue a node
            if node.left:
                queue.append(node.left)  # Enqueue the left child if it exists
            if node.right:
                queue.append(node.right)  # Enqueue the right child if it exists
    return depth  # Return the final depth

# Approach 6: Counting Nodes and Deriving Depth
def max_depth_count_nodes(root: Optional[TreeNode]) -> int:
    """
    Calculates the maximum depth of a binary tree by counting the total number of nodes
    and then estimating the depth based on that count.  This approach provides an
    approximation and is not accurate for unbalanced trees.  It's included to demonstrate
    a less common, and generally less reliable, method.

    Args:
        root (Optional[TreeNode]): The root of the binary tree.

    Returns:
        int: An *estimated* maximum depth of the tree.  This may not be accurate.
    """
    def count_nodes(node: Optional[TreeNode]) -> int:
        """
        Helper function to recursively count the number of nodes in a subtree.

        Args:
            node (Optional[TreeNode]): The root of the subtree.

        Returns:
            int: The number of nodes in the subtree.
        """
        if not node:
            return 0  # Base case: if the node is None, the count is 0
        # Recursively count the nodes in the left and right subtrees, and add 1 for the current node
        return 1 + count_nodes(node.left) + count_nodes(node.right)

    node_count = count_nodes(root)  # Get the total number of nodes in the tree
    # Estimate the depth.  For a perfectly balanced binary tree, the number of nodes is 2^depth - 1.
    # This formula is rearranged and approximated.  It's not accurate for unbalanced trees.
    estimated_depth = int((2 * node_count) ** 0.5)
    return estimated_depth  # Return the estimated depth
# Sample Usage
tree = TreeNode(1, TreeNode(2, TreeNode(4), TreeNode(5)), TreeNode(3, TreeNode(6), TreeNode(7)))
print("Depth with Time Delay:", max_depth_with_delay(tree))
print("Depth with Pickle Serialization:", max_depth_pickle(tree))
print("Depth with Multi-threading:", max_depth_multithreaded(tree))
print("Depth with JSON Serialization:", max_depth_json(tree))
print("Depth with BFS:", max_depth_bfs(tree))
print("Depth by Counting Nodes:", max_depth_count_nodes(tree))

File: Java/7-Trees/Problems/2-MaximumDepth-of-BinaryTree/MaximumDepth-of-BinaryTreePro.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;


public class MaximumDepthBinaryTree {

    /**
     * Approach 1: Recursive Depth-First Search (DFS)
     *
     * This is the classic, most intuitive approach.  It recursively explores the left and right subtrees
     * and returns the maximum depth.
     *
     * Time Complexity: O(N), where N is the number of nodes in the tree.
     * Space Complexity: O(H), where H is the height of the tree.  In the worst case (skewed tree), H = N,
     * and in the best case (balanced tree), H = log N.  This is due to the call stack.
     *
     * Real-world considerations:
     * -  Simple and easy to understand, making it good for maintainability.
     * -  Can lead to stack overflow for very deep trees, so it might not be suitable for extremely large datasets
     * without modification (see iterative approaches).
     */
    public int maxDepthRecursive(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int leftDepth = maxDepthRecursive(root.left);
        int rightDepth = maxDepthRecursive(root.right);
        return Math.max(leftDepth, rightDepth) + 1;
    }

    /**
     * Approach 2: Iterative Breadth-First Search (BFS)
     *
     * This approach uses a queue to perform a level-order traversal.  The depth is incremented at each level.
     *
     * Time Complexity: O(N), where N is the number of nodes in the tree.
     * Space Complexity: O(W), where W is the maximum width of the tree.  In the worst case (complete binary tree),
     * W = N/2, so the space complexity is O(N).  In the best case (skewed tree), W = 1, so
     * the space complexity is O(1).
     *
     * Real-world considerations:
     * -  Avoids stack overflow issues of the recursive approach.
     * -  Generally preferred for very large trees.
     * -  Slightly more complex to code than the recursive approach.
     */
    public int maxDepthBFS(TreeNode root) {
        if (root == null) {
            return 0;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int depth = 0;
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }
            depth++;
        }
        return depth;
    }

    /**
     * Approach 3: Iterative Depth-First Search (DFS) with Stack
     *
     * This approach uses a stack to simulate the recursive calls of the DFS approach.  It keeps track of the
     * depth of each node as it traverses the tree.
     *
     * Time Complexity: O(N), where N is the number of nodes in the tree.
     * Space Complexity: O(H), where H is the height of the tree.  In the worst case (skewed tree), H = N, and
     * in the best case (balanced tree), H = log N.
     *
     * Real-world considerations:
     * -  Avoids stack overflow issues of the recursive approach.
     * -  More memory-efficient than BFS in cases where the tree is deep but not very wide.
     * -  Can be slightly less intuitive than the recursive approach.
     */
    public int maxDepthIterativeDFS(TreeNode root) {
        if (root == null) {
            return 0;
        }
        Stack<Pair<TreeNode, Integer>> stack = new Stack<>();
        stack.push(new Pair<>(root, 1));
        int maxDepth = 0;
        while (!stack.isEmpty()) {
            Pair<TreeNode, Integer> pair = stack.pop();
            TreeNode node = pair.node;
            int depth = pair.depth;
            maxDepth = Math.max(maxDepth, depth);
            if (node.right != null) {
                stack.push(new Pair<>(node.right, depth + 1));
            }
            if (node.left != null) {
                stack.push(new Pair<>(node.left, depth + 1));
            }
        }
        return maxDepth;
    }

    // Helper class to store the node and its depth.  Using a generic Pair class for clarity.
    private static class Pair<T, U> {
        T node;
        U depth;
        Pair(T node, U depth) {
            this.node = node;
            this.depth = depth;
        }
    }

    /**
     * Approach 4:  Recursive DFS with Tail Recursion (Optimized - Not directly applicable in Java)
     *
     * Tail recursion is a specific form of recursion where the recursive call is the last operation
     * in the function.  Compilers can optimize tail recursion by turning it into iteration, thus avoiding
     * stack overflow.  However, Java does NOT perform tail call optimization.  This method is included for
     * conceptual completeness and for comparison with languages that do optimize tail recursion.
     *
     * Time Complexity: O(N)
     * Space Complexity: O(1) in languages that optimize tail recursion, O(H) in Java.
     *
     * Real-world considerations:
     * -  In Java, this offers no advantage over standard recursion.
     * -  In languages that *do* optimize tail recursion, this can be very efficient.
     */
    public int maxDepthTailRecursive(TreeNode root) {
        return maxDepthTailRecursiveHelper(root, 0);
    }

    private int maxDepthTailRecursiveHelper(TreeNode node, int depth) {
        if (node == null) {
            return depth;
        }
        int leftDepth = maxDepthTailRecursiveHelper(node.left, depth + 1);
        int rightDepth = maxDepthTailRecursiveHelper(node.right, depth + 1);
        return Math.max(leftDepth, rightDepth);
    }



    /**
     * Approach 5:  Level Order Traversal with Level Tracking (Similar to BFS, but with explicit level)
     *
     * This approach is very similar to BFS, but it explicitly tracks the level of each node as it's
     * being processed.  This makes the logic very clear for determining the maximum depth.
     *
     * Time Complexity: O(N)
     * Space Complexity: O(W), where W is the maximum width of the tree.
     *
     * Real-world considerations:
     * - Very clear and easy to understand.
     * - Avoids recursion, so it's suitable for large trees.
     * - Slightly more verbose than the basic BFS, but the added clarity can be beneficial in complex projects.
     */
    public int maxDepthLevelOrder(TreeNode root) {
        if (root == null) {
            return 0;
        }

        Queue<Pair<TreeNode, Integer>> queue = new LinkedList<>();
        queue.offer(new Pair<>(root, 1)); // Start at level 1
        int maxDepth = 0;

        while (!queue.isEmpty()) {
            Pair<TreeNode, Integer> current = queue.poll();
            TreeNode node = current.node;
            int level = current.depth;

            maxDepth = Math.max(maxDepth, level); // Update maxDepth

            if (node.left != null) {
                queue.offer(new Pair<>(node.left, level + 1));
            }
            if (node.right != null) {
                queue.offer(new Pair<>(node.right, level + 1));
            }
        }
        return maxDepth;
    }



    public static void main(String[] args) {
        // Create a sample binary tree for testing
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.right = new TreeNode(6);
        root.left.left.left = new TreeNode(7);


        MaximumDepthBinaryTree solution = new MaximumDepthBinaryTree();

        // Test the different approaches
        System.out.println("Max Depth (Recursive DFS): " + solution.maxDepthRecursive(root)); // Output: 4
        System.out.println("Max Depth (Iterative BFS): " + solution.maxDepthBFS(root));     // Output: 4
        System.out.println("Max Depth (Iterative DFS): " + solution.maxDepthIterativeDFS(root));  // Output: 4
        System.out.println("Max Depth (Tail Recursive): " + solution.maxDepthTailRecursive(root)); // Output: 4
        System.out.println("Max Depth (Level Order): " + solution.maxDepthLevelOrder(root));       // Output: 4
    }
}

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

